### ISSUE TITLE
Circular dependency error when using wildcard

### ISSUE THREAD

**Issue by HexSource** [PHP-DI/PHP-DI#379]
I'm not sure if I'm simply using features improperly but I don't understand why this generates an error.
This is a piece of my definitions file:
```
'Repository.*' => DI\factory(RepositoryFactory::class),
'Service.*.*' => function(DI\Container $c, \DI\Factory\RequestedEntry $entry){
    $serviceNS = "Application\\Service";
    $parts = explode('.',$entry->getName());
    $model = $parts[1];
    $service = $parts[2];

    $repository = $c->get("Repository.{$model}");
    $serviceClass = "{$serviceNS}\\{$model}\\{$service}";
    return new $serviceClass($repository);
},
```
I get a Circular dependency exception when I get to the line `$repository = $c->get("Repository.{$model}");`. Apparently  "Service.\*.\*" is identified as one of the candidates for "Repository.*", causing the exception.
Is this expected behavior? Am I doing something wrong?

**Comment by mnapoli** [comment]
It doesn't seem like you are doing something wrong :confused: By any chance do the `RepositoryFactory` get an entry that matches `Service.*.*`? (could you maybe post this class here?)

**Comment by mnapoli** [comment]
And maybe you could post a stack trace?  And if you have the time maybe a short case to reproduce it? (I know this one is more work, maybe we can figure it out from the code of `RepositoryFactory` and the stack trace first)

**Comment by HexSource** [comment]
The `RepositoryFactory` __invoke method is never actually called, so I don't think there's a problem with it. I reduced the case to a few lines and I was able to get the same error, so hopefully you can reproduce it too. The stack trace is also there, let me know if you need more details.

https://gist.github.com/HexSource/fc1257bfdd358ca6c36a

**Comment by HexSource** [comment]
Oh I figured it out. The problem is that my RepositoryFactory class is *not* a callable, but that's not the only thing. Here's what happens:

Take this definition: `'Repository.*' => DI\factory(RepositoryFactory::class),`

This is registered without an issue. Then you call, say, `$container->get('Repository.User')`

DI finds the match `'Repository.*'`, then starts processing `DI\factory(RepositoryFactory::class)`. PHP-DI is expecting a callable, so it goes searching for it. It figures that 'RepositoryFactory::class' is not a callable (my bad), so it goes looking for a key in the container that matches the string. Now here's is where things get fun.

At `CallableResolver`,  line 76 it starts looking for the key `RepositoryFactory`. We eventually get to `DefinitionArray->getDefinition`. At line 76, there's a `preg_match`. That matches each key in the container with the name we are looking for...now, remember that our key was `Repository.*`? Yeah, that dot is **not** escaped with `$key = addslashes($key);`, which causes the `RepositoryFactory` string to match. When it finds that  `'Repository.*' matches with `RepositoryFactory`, he starts the whole proccess again, and there's our circular dependency.

At the end of the day the mistake was mine for not making the factory callable, but I also got quite unlucky that my class name matched with a regex for a container's key.

So that's that.

**Comment by HexSource** [comment]
...that's not that. I tried fixing the callable thing and also using alternative methods (`DI\factory([FooFactory::class, 'create'])`and stuff), but I get the same error. I know it's definitely related to the dot and the regex, because if I simply replace the dot for something else it works fine, but I'm not sure how to make it work with the dot without potentially breaking other stuff.

**Comment by mnapoli** [comment]
Oh good catch with the not escaped dot. If you escape the `.` here: https://github.com/PHP-DI/PHP-DI/blob/master/src/DI/Definition/Source/DefinitionArray.php#L74 (e.g. with a `str_replace`) does that fix the problem?

**Comment by mnapoli** [comment]
I'll try to have a look at it today by the way.

**Comment by avant1** [comment]
You could use [preg_quote](http://php.net/manual/en/function.preg-quote.php) instead of combination of `addslashes` and `str_replace` to escape string for usage in regex.

**Comment by mnapoli** [comment]
@avant1 yep definitely better ;) (I just suggested str_replace for a quick and easy test :p)

**Comment by HexSource** [comment]
Yeah, it does fix it. It also works without changes to DI code if I escape my own stuff (`'Repository\.*' => DI\factory(RepositoryFactory::class),`, `$container->get('Repository\.User')`)

Also, `preg_quote` might be the direction to go, but I wonder if it's worth it.

If we just throw a `preg_quote` in every key, it means that users won't be able to make theirs keys regexes. I don't even know if this is a expected use case, but regardless this decision is bigger than just "make this dot work".

But I guess that's a decision for the author to make ;)

**Comment by mnapoli** [comment]
Regexes are not supported in container keys, it's not a feature (and it's not documented, only `*` is allowed) so we can use `preg_quote`.

Thanks for the trying and letting us know, if anyone wants to create a PR I'd happily merge it, else I'll get to it tonight.

**Comment by mnapoli** [comment]
I've released [5.2.2](https://github.com/PHP-DI/PHP-DI/releases/tag/5.2.2)
